File: python/socketd/test/labs/SubProtocolsTest.py
import asyncio

from socketd import SocketD
from socketd.transport.core.Message import Message
from socketd.transport.core.Session import Session
from socketd.transport.core.entity.StringEntity import StringEntity
from socketd.transport.core.listener.SimpleListener import SimpleListener
from socketd.utils.LogConfig import log


class ListenerImpl(SimpleListener):
    async def on_open(self, s: Session):
        # 会话打开时
        log.info(s.session_id())
        pass
    async def on_message(self, s: Session, m: Message):
        # 收到任意消息时（方便做统一的日志打印）
        log.info(m.data_as_string())


async def main():
    await (SocketD.create_server("sd:ws")
                    .config(lambda c: c.port(8602).use_subprotocols(True))
                    .listen(ListenerImpl()).start())

    session = await (SocketD.create_client("sd:ws://localhost:8602")
                     .config(lambda c: c.use_subprotocols(False))
                     .open_or_throw())

    session.send("/demo", StringEntity("hello"))

    await asyncio.Future()


if __name__ == "__main__":
    asyncio.run(main())

File: java/socketd-transport-test/src/test/java/labs/SubprotocolsTest.java
package labs;

import org.noear.socketd.SocketD;
import org.noear.socketd.transport.client.ClientSession;
import org.noear.socketd.transport.core.entity.StringEntity;
import org.noear.socketd.transport.core.listener.EventListener;

/**
 * @author noear 2024/7/21 created
 */
public class SubprotocolsTest {
    static final String[] schemas = new String[]{
            "sd:ws-java",
    };

    public static void main(String[] args) throws Exception {
        SocketD.createServer(schemas[0])
                .config(c -> c.port(8602).useSubprotocols(true))
                .listen(new EventListener().doOnOpen(s -> {
                    System.out.println("..................: " + s.sessionId());
                }).doOnMessage((s, m) -> {
                    System.out.println("..................: " + m.dataAsString());
                }))
                .start();

        String serverUrl = schemas[0] + "://127.0.0.1:8602/path?u=a&p=2";

        ClientSession session = SocketD.createClient(serverUrl)
                .config(c -> c.useSubprotocols(true))
                .openOrThow();
        session.send("/demo", new StringEntity("hello"));
    }
}